File: leetcode/ayaan/week4/SearchInsertPosition.java
package leetcode.ayaan.week4;

public class SearchInsertPosition {
    public static void main(String[] args) {
        int[] nums = {3,6,7,8,10};
        int target = 5;

        int result = searchInsert(nums, target);
        System.out.println(result);
    }

    public static int searchInsert(int[] nums, int target) {
        int result = 0;
        for(int i=0; i<nums.length; i++){
            int num = nums[i];
            if(num == target){
                result = i;
                break;
            } else if (num > target){
                if(i == 0){
                    if(target <= 0){
                        result = 0;
                    } else {
                        result = target-1;
                    }
                } else {
                    result = (i-1) + (target - nums[i-1]);
                }
                break;
            }

            if(i == nums.length-1){
                result = i + (target - nums[i]);
            }
        }
        return result;
    }
}

File: leetcode/Hyun/week4/search_insert_position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        size = len(nums)
        left = 0
        right = size - 1

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                if mid > 0 and target > nums[mid - 1]:
                    return mid
                if mid == 0:
                    return 0
                right = mid - 1
            else:  # nums[mid] < target
                if mid + 1 < size and target < nums[mid + 1]:
                    return mid + 1
                if mid == size-1:
                    return size
                left = mid + 1